problem/views.py

```diff
@@ -65,10 +64,17 @@ def problem(request):
                    'can_add_problem': can_add_problem})
 
 def detail(request, pid):
-    user = request.user
+    user = validate_user(request.user)
     tag_form = TagForm()
     try:
         problem = Problem.objects.get(pk=pid)
+        last_contest = problem.contest_set.all().order_by('-start_time')
+        if len(last_contest) and last_contest[0].start_time < timezone.now():
```

Comment:
I prefer to put this checking in ```if not has_problem_auth```, so that the server won't check it every time.
(Only invisible problems need this checking)